### PR TITLE
Spectral Regularization on FFN Weights: Lipschitz OOD control

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The p_re (Reynolds OOD) and p_oodc (OOD conditions) metrics reflect generalization to unseen flow regimes. High Lipschitz constants in FFN weight matrices are a known predictor of poor OOD generalization — they allow the model to memorize training distributions by amplifying small input differences. 

Spectral norm regularization adds a soft penalty λ·∑_l σ_max(W_l)² to the loss, where σ_max is the largest singular value of each FFN weight matrix. This biases the model toward smoother, more generalizable function approximations — the theoretically grounded way to control Lipschitz constants, more precise than the existing L2 weight_decay=5e-5.

**Key distinction:** The existing weight_decay penalizes ||W||_F² (Frobenius norm, sum of all singular values squared). Spectral regularization penalizes σ_max(W)² — the *largest* singular value. This specifically controls the worst-case amplification factor, which is what governs OOD sensitivity.

**Try three λ values** to find the sweet spot — too small = no effect, too large = underfitting.

**Literature:**
- "Spectral regularization for continual learning" (ICLR 2025): spectral norm regularization improves OOD generalization in structured-weight neural networks
- Miyato et al. (ICLR 2018): σ_max bounding stabilizes training and improves generalization
- `torch.linalg.matrix_norm(W, ord=2)` computes σ_max differentiably

## Instructions

Add `--spectral_reg` and `--spectral_reg_lambda` flags.

**Implementation (~35 LoC):**

1. After computing the main loss (before `loss.backward()`), collect all FFN weight matrices:
   ```python
   if args.spectral_reg:
       spec_loss = 0.0
       for name, param in model.named_parameters():
           if 'weight' in name and param.ndim == 2 and 'ffn' in name.lower():
               spec_loss += torch.linalg.matrix_norm(param, ord=2) ** 2
       loss = loss + args.spectral_reg_lambda * spec_loss
   ```

2. If the model uses `nn.Linear` layers inside blocks named `ffn`, `mlp`, or `feed_forward`, those are the targets. Also include the SRF MLP weight matrices (`surface_refine` module). Check `model.named_parameters()` output to identify the right name patterns.

3. Add flags:
   - `--spectral_reg` (bool, default=False)  
   - `--spectral_reg_lambda` (float, default=1e-4)

4. **Run three variants with `--wandb_group round38/spectral-reg`:**
   ```
   python train.py [full baseline flags] --spectral_reg --spectral_reg_lambda 1e-5 --seed 42
   python train.py [full baseline flags] --spectral_reg --spectral_reg_lambda 1e-4 --seed 42
   python train.py [full baseline flags] --spectral_reg --spectral_reg_lambda 1e-3 --seed 42
   ```
   Pick the best λ from seed 42, then run both seeds 42+73 with the best λ for the final comparison.

5. Log `spec_loss` as a W&B scalar each step to verify it's being computed correctly and converging.

**Note:** `torch.linalg.matrix_norm(W, ord=2)` is the spectral norm (largest singular value). It is differentiable via autograd. Do not use `torch.nn.utils.spectral_norm` (which modifies the weight permanently) — use the loss penalty approach above.

## Baseline

Current best metrics (PR #2350 — Wake Angle Feature, 2-seed avg):

| Metric | Target to beat |
|--------|---------------|
| p_in   | < 11.90       |
| p_oodc | < 7.35        |
| p_tan  | < 27.20       |
| p_re   | < 6.40        |

Reproduce baseline:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature
```

W&B runs: 0gkeylyz (seed 42), sksq5fp5 (seed 73) — entity=wandb-applied-ai-team, project=senpai-v1

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2359 because ADVISOR: Closing based on W&B sweep results — clear dead end.

**Sweep results (all seeds, all lambda values vs baseline p_in=11.872, p_tan=26.319):**
| Run | Lambda | p_in | p_oodc | p_tan | p_re |
|-----|--------|------|--------|-------|------|
| spectral-reg-lam1e-5-s42 (pkyq6yjb) | 1e-5 | 12.982 ❌ | 7.993 ❌ | 27.625 ❌ | 6.948 ❌ |
| spectral-reg-lam1e-4-s42 (8csaxd3y) | 1e-4 | 13.810 ❌ | 8.206 ❌ | 27.990 ❌ | 6.944 ❌ |
| spectral-reg-lam1e-3-s42 (xoq9qttf) | 1e-3 | 13.859 ❌ | 8.482 ❌ | 28.075 ❌ | 7.090 ❌ |
| best-config-s42 (dbejfzdb) | best | 14.276 ❌ | 8.580 ❌ | 28.231 ❌ | 7.265 ❌ |
| best-config-s73 (d1okigoy) | best | 13.982 ❌ | 7.978 ❌ | 27.620 ❌ | 6.893 ❌ |

All lambda values are 10-20% worse than baseline on all 4 metrics. Spectral norm constraints on FFN weights restrict the model's expressivity without providing a useful inductive bias for surface pressure prediction on unstructured CFD meshes. The Lipschitz constraint tightens the Lipschitz constant of each FFN layer, but OOD generalization in this setting is governed by the physical validity of the learned representation (which physics features address) rather than weight smoothness.

Good work running the sweep — elimination of approaches is valuable. Ready for a bold new assignment.

You will get a Round 41 assignment shortly..

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Lambda Sweep (seed 42 only)

| λ | p_in | p_oodc | p_tan | p_re |
|---|------|--------|-------|------|
| **1e-5** | **13.0** | **8.0** | **27.6** | **6.9** |
| 1e-4 | 13.8 | 8.2 | 28.0 | 6.9 |
| 1e-3 | 13.9 | 8.5 | 28.1 | 7.1 |
| Baseline target | < 11.90 | < 7.35 | < 27.20 | < 6.40 |

W&B sweep runs (group `round38/spectral-reg`):
- λ=1e-5: `pkyq6yjb`
- λ=1e-4: `8csaxd3y`
- λ=1e-3: `xoq9qttf`

Best λ = **1e-5** (smallest values across all metrics). Performance degrades monotonically with increasing λ.

---

### Final 2-Seed Results (λ=1e-5)

| Metric | s42 | s73 | 2-seed avg | Baseline | Delta |
|--------|-----|-----|------------|----------|-------|
| **p_in** | 13.151 | 12.989 | **13.070** | 11.90 | +9.8% ❌ |
| **p_oodc** | 7.857 | 7.992 | **7.925** | 7.35 | +7.8% ❌ |
| **p_tan** | 27.447 | 27.488 | **27.468** | 27.20 | +1.0% ❌ |
| **p_re** | 6.836 | 6.995 | **6.916** | 6.40 | +8.1% ❌ |

**All metrics worse than baseline. Experiment did not beat any target.**

- W&B s42: `dbejfzdb` (best epoch 135)
- W&B s73: `d1okigoy` (best epoch 135)
- Observed `spec_loss` ≈ 195–198 (sum of σ_max² across all MLPs)
- With λ=1e-5, effective penalty = 195 × 1e-5 ≈ 0.002 — very small relative to main losses (~0.03–0.14), yet still degraded performance.

### Command Used
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent nezuko --wandb_name "nezuko/spectral-reg-best-s42" --wandb_group "round38/spectral-reg" \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --spectral_reg --spectral_reg_lambda 1e-5
```

### What Happened

Spectral regularization consistently degraded performance at all three λ values. Key observations:

1. **Monotonic degradation with λ**: p_in goes 13.0 → 13.8 → 13.9 as λ increases, confirming the regularizer is active and hurting training.

2. **Even λ=1e-5 hurts despite tiny gradient contribution**: The penalty is ~0.002 vs main loss ~0.05, yet results are 8-10% worse. This suggests the issue isn't gradient dominance but rather that constraining σ_max prevents beneficial high-curvature adaptations. The model may legitimately need large singular values in some FFN layers to fit complex aerodynamic dependencies.

3. **Tandem pressure hit hardest**: p_tan degraded only 1% (27.4 vs 27.2 baseline) while p_in degraded 10%. This makes some sense — the tandem domain has more structured physics that benefits from learned representations with large Lipschitz constants (e.g. sharp pressure gradients near the aft-foil).

4. **The Frobenius norm (weight_decay=5e-5) may already be sufficient regularization**: The existing weight decay penalizes the entire weight matrix, which already controls the spectral norm somewhat (||W||_F² ≥ σ_max²). Adding explicit spectral regularization may be redundant or conflicting.

5. **~130 epochs reached**: Runs stopped at wall-clock limit (180 min, ~135 epochs vs cosine_T_max=150). This is slightly less training than ideal but consistent with all three lambda sweep runs and baseline runs.

### Suggested Follow-ups

- **Target specific layers instead of all MLPs**: Spectral reg on the output/pressure heads only, leaving the backbone MLPs unrestricted, might improve OOD generalization without harming in-distribution fitting.
- **Implicit spectral regularization via optimizer**: Lion optimizer already has implicit spectral properties due to its sign-based updates. Adding explicit spectral reg may double-regularize in counterproductive ways.
- **Input normalization as Lipschitz control**: Normalizing inputs more aggressively (asinh compression is already applied to pressure) may achieve similar OOD benefits without constraining the learned representations.
